Feature/recover password

jamcunha opened the pull request:
Closes #84 

# Review checklist
-   [ ] Properly documents API changes in `docs/openapi.yml`
-   [ ] Contains enough appropriate tests
-   [ ] Behavior is as expected
-   [ ] Clean, well structured code

jamcunha committed:
added recover request route

jamcunha committed:
added password update route

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -27,6 +28,13 @@ class AuthController(val authService: AuthService) {
+    fun generateRecoveryToken(@PathVariable id: Long): Map<String, String> {
+        val recoveryToken = authService.generateRecoveryToken(id)
+        // TODO: Change URL Later
+        return mapOf("recovery_url" to "localhost:8080/accounts/recoverPassword/$recoveryToken")
```
We'll probably need to set some property for the URL

BrunoRosendo reviewed (changes requested):
The logic looks good overall. I didn't do a careful review since this is just a draft

jamcunha committed:
added tests

jamcunha committed:
added docs

jamcunha committed:
fix documentation

BrunoRosendo commented:
@jamcunha Is this ready for review or do you still have some doubts? We can talk about it in the next meeting or in 1on1

jamcunha commented on src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:
```diff
@@ -901,6 +904,80 @@ class AccountControllerTest @Autowired constructor(
+                        passwordRecoveryPayload,
+                        "Recover password",
+                        "Update the password of an account using a recovery token.",
+                        // urlParameters = parameters,
```
It is pretty much ready for review. There is a doubt here because it errors when I document the URL parameters.

bdmendes reviewed (commented):
The code looks good to me. To improve security a bit, we can look into serializing the current password into the jwt to make sure it is only used once: https://stackoverflow.com/a/54865104

jamcunha committed:
added a property for backend url

jamcunha committed:
Merge branch 'develop' into feature/recover-password

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/dto/auth/PassRecoveryDto.kt:
```diff
@@ -0,0 +1,5 @@
+package pt.up.fe.ni.website.backend.dto.auth
+
+data class PassRecoveryDto(
+    val password: String
```
I'd consider putting a constraint to enforce a minimum size to the password, as we have in the account model. Currently, I can reset to an easy password like `1`

jamcunha replied:
While doing this, I noticed that `ChangePasswordDto` has the same problem. Should I fix it in this PR?

BrunoRosendo replied:
Yes that would be great!

BrunoRosendo reviewed (changes requested):
Very good job with this implementation! I left some comments with small suggestions and doubts

jamcunha commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -28,6 +33,13 @@ class AuthController(val authService: AuthService) {
+    fun generateRecoveryToken(@PathVariable id: Long): Map<String, String> {
+        val recoveryToken = authService.generateRecoveryToken(id)
+        // TODO: Change URL Later
+        return mapOf("recovery_url" to "$backendUrl/accounts/recoverPassword/$recoveryToken")
```
I should rename that to `apiUrl` as it better describes my thoughts when implementing it. I didn't consider being a frontend page because we didn't have one, but certainly is the intention when implementing sending emails.

BrunoRosendo replied:
Is there any point leaving it with the backend URL? Why not implement the right logic and only send the email in the future issue?

jamcunha replied:
Makes sense. Do we already have the frontend page?

BrunoRosendo replied:
We don't but just put a config property as a placeholder

jamcunha committed:
review changes

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/dto/auth/ChangePasswordDto.kt:
```diff
@@ -1,6 +1,12 @@
+import pt.up.fe.ni.website.backend.model.constants.AccountConstants as Constants
+
 data class ChangePasswordDto(
+    @field:Size(min = Constants.Password.minSize, max = Constants.Password.maxSize)
```
Checking the old password seems redundant. It probably won't make a difference but this could block changing passwords if we ever populate the db directly (again, shouldn't happen but ***could*** happen)

jamcunha replied:
Makes sense, removed it.

BrunoRosendo reviewed (approved):
Good job, I left a small comment but I'm pre-approving. It'd be nice if you could review your commits too, maybe squash a few

BrunoRosendo commented on src/test/kotlin/pt/up/fe/ni/website/backend/controller/AccountControllerTest.kt:
```diff
@@ -901,6 +905,83 @@ class AccountControllerTest @Autowired constructor(
+                        documentRequestPayload = true
+                    )
+                }
+        }
```
@jamcunha please don't forget to create the issue :)

jamcunha replied:
Done. Please let me know if more information is needed.
Issue: #171

jamcunha committed:
refactor recover password to single use jwt

jamcunha commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -59,29 +59,43 @@ class AuthService(
-    private fun generateToken(account: Account, expiration: Duration, isRefresh: Boolean = false): String {
+    private fun generateToken(
+        account: Account,
+        expiration: Duration,
+        isRefresh: Boolean = false,
+        usePasswordHash: Boolean = false
+    ): String {
         val roles = if (isRefresh) emptyList() else getAuthorities() // TODO: Pass account to getAuthorities()
         val scope = roles
             .stream()
             .map(GrantedAuthority::getAuthority)
             .collect(Collectors.joining(" "))
         val currentInstant = Instant.now()
-        val claims = JwtClaimsSet
+        val claimsBuilder = JwtClaimsSet
             .builder()
             .issuer("self")
             .issuedAt(currentInstant)
             .expiresAt(currentInstant.plus(expiration))
             .subject(account.email)
             .claim("scope", scope)
-            .build()
+
+        if (usePasswordHash) {
+            claimsBuilder.claim("passwordHash", account.password)
+        }
+
+        val claims = claimsBuilder.build()
         return jwtEncoder.encode(JwtEncoderParameters.from(claims)).tokenValue
     }
```
I don't know if this is what to do regarding the password serialization mentioned in this PR, but it seems right to me.

DoStini replied:
Is this regarding password changing only 1 time with the link?

jamcunha replied:
Yes

BrunoRosendo reviewed (changes requested):
Regarding the latest change, where do you check to see if the hash present in the jwt is the same as the one in the db?

BrunoRosendo commented:
> Regarding the latest change, where do you check to see if the hash present in the jwt is the same as the one in the db?

nevermind this comment, I was looking at the wrong diff

BrunoRosendo reviewed (changes requested):
can you create 1/2 tests for the latest change? Maybe 1 where a user tries to use the same recovery link twice and another one just to check for missing hash (if the last one isn't too hard)

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/ErrorMessages.kt:
```diff
@@ -11,6 +11,12 @@ object ErrorMessages {
+
+    const val expiredRecoveryToken = "password recovery token has expired"
+
+    const val missingHashClaim = "password hash claim is missing"
```
I'm thinking if we should return a general `invalid recovery token` error instead of specifying it. This may be a detail we want to omit from users, wdyt @DoStini ?

jamcunha replied:
I did it that way because it would only error if it's not implemented well in the code (forgetting to set `usePasswordHash` to true and use it). But if it eventually occurs with the user, it makes sense to omit it.

DoStini replied:
I was just reading this now, I've mentioned in my previous comments that we should omit these kinds of errors to the users so that information on who uses the website doesn't get leaked. Always return ok even if the account doesn't exist

DoStini replied:
Now that I'm reading this again, it makes sense to have an `invalid recovery token` error whenever we try to set a new password.

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -57,26 +57,45 @@ class AuthService(
         return generateAccessToken(account)
     }
 
+    fun generateRecoveryToken(id: Long): String {
```
```suggestion
    private fun generateRecoveryToken(id: Long): String {
```

jamcunha replied:
If it's private it will not work, or am I missing something?

BrunoRosendo replied:
you're right of course, my bad

DoStini commented:
Overall looking good, just left some comments to improve the RESTiness of the API and privacy concerning issues 🚀

jamcunha commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -14,6 +16,9 @@ import pt.up.fe.ni.website.backend.service.AuthService
 @RestController
 @RequestMapping("/auth")
 class AuthController(val authService: AuthService) {
+    @field:Value("\${page.recover-password}")
+    private lateinit var recoverPasswordPage: String
+
```
To use the property in `application.properties` file.

bdmendes commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AccountController.kt:
```diff
@@ -29,8 +31,20 @@ class AccountController(private val service: AccountService) {
     fun getAccountById(@PathVariable id: Long) = service.getAccountById(id)
 
+    @PutMapping("/recoverPassword/{recoveryToken}")
+    fun recoverPassword(
+        @Valid @RequestBody
+        dto: PasswordRecoveryDto,
+        @PathVariable recoveryToken: String
+    ) =
+        service.recoverPassword(recoveryToken, dto)
+
     @PostMapping("/changePassword/{id}")
-    fun changePassword(@PathVariable id: Long, @RequestBody dto: ChangePasswordDto): Map<String, String> {
+    fun changePassword(
+        @Valid @RequestBody
+        dto: ChangePasswordDto,
+        @PathVariable id: Long
+    ): Map<String, String> {
```
Indeed, one can get the authenticated account via a method in the auth service

bdmendes commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AccountService.kt:
```diff
@@ -65,6 +70,29 @@ class AccountService(
+        }
+        val account = getAccountByEmail(jwt.subject)
+
+        val tokenPasswordHash = jwt.getClaim<String>("passwordHash")
+            ?: throw InvalidBearerTokenException(ErrorMessages.missingHashClaim)
```
What if the hash is not in the payload (since we have that option)? You can simply let it be null and skip the next check if that is the case.

jamcunha committed:
refactored endpoinnts to be email ready

jamcunha commented:
Just changed the recovery URL request to be provided with the email instead of the id and removed errors related to invalid tokens, logging them instead.

> Seems weird to me that this is in a different controller.
> 
> I suggest either moving everything to auth controller or to account controller.
> 
> Besides that, there are two requests with a similar URL `recoverPassword`.
> 
> I suggest the following routes in AuthController: post -> password/recovery with email in body post -> password/recovery/{token}/confirm post -> password/update

The tests may be lacking because I didn't want to test more without making these changes.

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/dto/auth/ChangePasswordDto.kt:
```diff
@@ -1,6 +1,11 @@
 data class ChangePasswordDto(
     val oldPassword: String,
+
+    @field:Size(min = Constants.Password.minSize, max = Constants.Password.maxSize)
```
@DoStini the maximum size is also used when creating an account. I believe this was done to prevent ridiculously long passwords but, now that I think about it, we only store the hash so there's no point. I believe it's fine to use the default value of Integer.MAX_VALUE here and in the account model

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AccountService.kt:
```diff
@@ -65,6 +70,29 @@ class AccountService(
+            try {
+                jwtDecoder.decode(recoveryToken)
+            } catch (e: Exception) {
+                throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
```
@DoStini do we always assume a password change was successful even if the URL is malformated? It seems weird for the user to think the password changed when reality it didn't

DoStini replied:
You are right, I missunderstood the use of this function. Only errors in the act of `requesting password change` should be hidden from the user. If the URL is malformed or expired whenever they are setting the new password, the user should be warned about it!

This should go back to how it was. Sorry @jamcunha !!!

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AccountService.kt:
```diff
@@ -17,8 +19,9 @@ import pt.up.fe.ni.website.backend.utils.extensions.filenameExtension
+    private val jwtDecoder: JwtDecoder,
     private val fileUploader: FileUploader
-) {
+) : Logging {
```
Is the logger being used?

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -57,26 +59,84 @@ class AuthService(
+            try {
+                jwt.getClaim<String>("passwordHash")
+            } catch (e: Exception) {
+                // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
```
please remove these comments when it's done

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -57,26 +59,84 @@ class AuthService(
+                jwtDecoder.decode(recoveryToken)
+            } catch (e: Exception) {
+                // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
+                logger.error(ErrorMessages.invalidRecoveryToken)
```
I don't think error makes sense here. logger.errors is usually for server-side or unexpected errors, this is just a malformed request. Imo info or warn makes more sense

BrunoRosendo commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -57,26 +59,84 @@ class AuthService(
+        if (account.password != tokenPasswordHash) {
+            // throw InvalidBearerTokenException(ErrorMessages.invalidRecoveryToken)
+            logger.error(ErrorMessages.invalidRecoveryToken)
+            return null
+        }
+
+        account.password = passwordEncoder.encode(dto.password)
+        return accountService.updateAccount(account)
```
@DoStini this is related to my previous comment but, if we want to hide invalid tokens then the user can distinguish between null and the account object. In that case, we should always return empty map on success

jamcunha replied:
But if we remove that, how can we tell if it is or isn't successful?

BrunoRosendo reviewed (changes requested):
small changes, Im also waiting for @DoStini response on my comment

DoStini commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -14,6 +16,9 @@ import pt.up.fe.ni.website.backend.service.AuthService
 @RestController
 @RequestMapping("/auth")
 class AuthController(val authService: AuthService) {
+    @field:Value("\${page.recover-password}")
+    private lateinit var recoverPasswordPage: String
```
Shouldn't this be passed by config and config properties or is it bloat to use? @BrunoRosendo @bdmendes

BrunoRosendo replied:
What do you mean? It is passed by the config property

BrunoRosendo reviewed (changes requested):
Small comments, besides the one still unresolved

Also, can you find a way to test the uncovered (red) lines? It'd be awesome if it's not too much trouble

![image](https://github.com/NIAEFEUP/website-niaefeup-backend/assets/55768934/8693c337-8780-4e10-ac27-9fda18aafd34)

jamcunha commented on src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt:
```diff
@@ -180,6 +193,268 @@ class AuthControllerTest @Autowired constructor(
+        }
+
+        @Test
+        fun `should fail if email is not found`() {
```
It should. I mistakenly removed it after reading something about failing when confirming the token and assumed fail all for some reason.

jamcunha commented:
> Small comments, besides the one still unresolved
> 
> Also, can you find a way to test the uncovered (red) lines? It'd be awesome if it's not too much trouble
> 
> ![image](https://user-images.githubusercontent.com/55768934/254106136-8693c337-8780-4e10-ac27-9fda18aafd34.png)

I'm having problems trying to test those two lines because when testing token expiration, I try to create a new token with all the claims of the previous token but modifying the `expireAt` to `Instant.now()` but it's caught in an exception when trying to decode in `confirmRecoveryToken`. I think it's the same for `passwordHash` claim.

jamcunha committed:
changed endpoint error handling and minor fixes

jamcunha committed:
improved jwt token error handling

jamcunha commented:
For now, I only changed the error handling in the recovery logic so as not to mess up with the tests. After some inputs, I will change the handling for other jwt logic.

BrunoRosendo reviewed (approved):
Needs a rebase but I'm pre-approving

LuisDuarte1 reviewed (approved):
LGTM 🚀, it looks clean to me, just solve the conflicts below 😅

LuisDuarte1 commented:
@jamcunha can you fix the conflicts please?

jamcunha committed:
Merge branch 'develop' into feature/recover-password

DoStini commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -28,6 +36,21 @@ class AuthController(val authService: AuthService) {
 
+    @PostMapping("/password/recovery")
+    fun generateRecoveryToken(@RequestBody recoveryRequestDto: PasswordRecoveryRequestDto): Map<String, String> {
+        val recoveryToken = authService.generateRecoveryToken(recoveryRequestDto.email)
+            ?: return emptyMap()
+        // TODO: Change to email service
+        return mapOf("recovery_url" to "$recoverPasswordPage/$recoveryToken/confirm")
```
Make this return OK all the time and comment on what is supposed to go on an email. If someone wants to test this right away, one can simply print the result. It is safer to leave the correct behavior already implemented.

```suggestion
        val recoveryToken = authService.generateRecoveryToken(recoveryRequestDto.email)
            ?: return emptyMap()
        // TODO:  Add email service with payload "$recoverPasswordPage/$recoveryToken/
        
        return ""
```

DoStini replied:
I'm not sure about the `return ""`, but it should always be status 200. 

Also document on why this is done for security reasons, in order to not leak what are the emails associated with our service

jamcunha replied:
Couldn't we have it return the recovery link for now so we can test the feature and remove it when the email service is complete?

If I'm not wrong it always returns status 200.

DoStini commented on src/test/kotlin/pt/up/fe/ni/website/backend/controller/AuthControllerTest.kt:
```diff
@@ -180,6 +201,367 @@ class AuthControllerTest @Autowired constructor(
+        }
+
+        @Test
+        fun `should return empty if email is not found`() {
+            mockMvc.perform(
+                post("/auth/password/recovery")
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(objectMapper.writeValueAsString(mapOf("email" to "dont@exist.com")))
+            )
+                .andExpectAll(
+                    status().isOk(),
+                    jsonPath("$.recovery_url").doesNotExist()
+                )
+                .andDocument(
+                    documentation,
+                    "Recover password",
+                    "This endpoint operation allows the recovery of the password of an account, " +
+                        "sending an email with a link to reset the password.",
+                    documentRequestPayload = true
+                )
+        }
+
+        @Test
+        fun `should return password recovery link`() {
+            mockMvc.perform(
+                post("/auth/password/recovery")
+                    .contentType(MediaType.APPLICATION_JSON)
+                    .content(objectMapper.writeValueAsString(mapOf("email" to testAccount.email)))
+            )
+                .andExpectAll(
+                    status().isOk(),
+                    jsonPath("$.recovery_url").exists()
+                ).andDocument(
+                    documentation,
+                    "Recover password",
+                    "This endpoint operation allows the recovery of the password of an account, " +
+                        "sending an email with a link to reset the password.",
+                    documentRequestPayload = true
+                )
+        }
```
As I explained before, this is not the expected behavior. Both cases will return empty and ok status.

What you should do is test the request returns empty and ok both times, but also to test if the first situation does not call the generate token function. I think the easiest way would be by mocking JwtClaimsSet and verifying if some of those functions do not get called.

jamcunha replied:
Same as above, this could be as is for now and changed when the email service is implemented.
Can leave a comment to remind that.

DoStini commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/AuthController.kt:
```diff
@@ -28,6 +36,21 @@ class AuthController(val authService: AuthService) {
 
+    @PostMapping("/password/recovery")
+    fun generateRecoveryToken(@RequestBody recoveryRequestDto: PasswordRecoveryRequestDto): Map<String, String> {
+        val recoveryToken = authService.generateRecoveryToken(recoveryRequestDto.email)
+            ?: return emptyMap()
```
```suggestion
        val recoveryToken = authService.generateRecoveryToken(recoveryRequestDto.email)
        
```

rubuy-74 replied:
Not needed as this is resolved in the comment below

jamcunha commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -44,39 +45,69 @@ class AuthService(
-            }
-        if (jwt.expiresAt?.isBefore(Instant.now()) != false) {
-            throw InvalidBearerTokenException(ErrorMessages.expiredRefreshToken)
-        }
+        val jwt = jwtDecoder.decode(refreshToken)
```
I removed the try catch block and handled the exception in the  `ErrorController` and the token expiration validation is already handled in the `decode()` method.

jamcunha commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -44,39 +45,69 @@ class AuthService(
+        val account = try {
+            accountService.getAccountByEmail(email)
+        } catch (e: Exception) {
+            return null
```
If someone uses an email that's not related to an account it would trigger the exception. I don't think we should log this.

rubuy-74 committed:
feat:requested changes

rubuy-74 committed:
Merge branch 'develop' into feature/recover-password

rubuy-74 committed:
fix:lint

DoStini reviewed (approved):
Honestly I dont even know what I requested to change but if my homie @rubuy-74 asks me to allow the merge of this beautiful pull request, I will. I hope this project finished this year.

This will be my last LGTM 🚀 A hug to all the beautiful members of NIAEFEUP.

rubuy-74 committed:
Merge branch 'develop' into feature/recover-password

MRita443 reviewed (changes requested):
Looks great, some small details and conversations I unresolved. Also when attempting to change password and giving a password too short, it returns the attempted password back to the user, maybe that's not a good idea

rubuy-74 commented on src/main/kotlin/pt/up/fe/ni/website/backend/service/AuthService.kt:
```diff
@@ -44,39 +45,69 @@ class AuthService(
+
+        account.password = passwordEncoder.encode(dto.password)
+        return accountService.updateAccount(account)
+    }
```
This is already resolved. I separated the updating password section to a different function that is called if the confirmRecoryToken function succeds

rubuy-74 committed:
feat:requested changes

rubuy-74 committed:
fix:fix null-safety error

rubuy-74 commented on src/main/kotlin/pt/up/fe/ni/website/backend/controller/ErrorController.kt:
```diff
@@ -141,6 +144,22 @@ class ErrorController(private val objectMapper: ObjectMapper) : ErrorController,
+    @ExceptionHandler(JwtValidationException::class)
+    @ResponseStatus(HttpStatus.UNAUTHORIZED)
+    fun invalidAuthentication(e: JwtValidationException): CustomError {
+        return if (e.message?.contains("expired") == true) {
```
This can't be because e.message?.contains("expired") return a Boolean? and the expression must return a Boolean. The "== true" is basically forcing it to be a valid Boolean and not null

MRita443 committed:
Merge branch 'develop' into feature/recover-password

MRita443 reviewed (approved):
LGTM
